luarules/gadgets/unit_custom_weapons_overpen.lua

```diff
@@ -0,0 +1,449 @@
+local function overpenDamage(unitID, unitDefID, featureID, damage, projID, attackID)
+    local driver = drivers[projID]
+    drivers[projID] = nil
+
+    local weaponData = driver[1]
+    local damageLeft = driver[2]
+
+    local health, healthMax, damageType
+    if unitID then
+        health, healthMax = spGetUnitHealth(unitID)
+        damageType = weaponData.damages[unitArmorType[unitDefID]]
+    elseif featureID then
+        health, healthMax = Spring.GetFeatureHealth(unitID)
+        damageType = weaponData.damages[0]
+    end
+    local damageBase = min(damage, damageType)
+
+    if weaponData.falloff then
+        damage = damage * damageLeft
+        damageBase = damageBase * damageLeft
+        damageLeft = damageLeft - health / damageBase - weaponData.penalty
+
+        if weaponData.expDefID and damageLeft <= explodeThreshold then
+            explodeDriver(projID, weaponData.expDefID, attackID, unitID, featureID)
+            return damage
+        end
+    end
+
+    if damageBase > health and damageBase >= healthMax * damageThreshold then
+        if damageLeft > 0 then
+            if weaponData.falloff then
+                driver[2] = damageLeft
+            end
+            respawn[projID] = driver
+        end
+        return damage
+    elseif weaponData.expDefID then
+        explodeDriver(projID, weaponData.expDefID, attackID, unitID, featureID)
+        return damage
+    end
+
+    local damageDone = min(1, damageBase / damageType)
+    return damage, inertia(damageDone) * damageDone
+end
+
+---Simulate the overpen effect by creating a new projectile.
+local function spawnPenetrator(projID, attackID, penDefID, unitID, featureID)
+    local penetrator = respawn[projID]
+    respawn[projID] = nil
+
+    local px, py, pz = spGetProjectilePosition(projID)
+    local timeToLive = spGetProjectileTimeToLive(projID)
+    local vx, vy, vz, vw = spGetProjectileVelocity(projID)
+
+    local driverData = penetrator[1]
+    local explodeID = driverData.expDefID
+
+    if driverData.slowing ~= nil then
+        local speedLeft = inertia(penetrator[2])
+        vx, vy, vz, vw = vx * speedLeft,
+                         vy * speedLeft,
+                         vz * speedLeft,
+                         vw * speedLeft
+    end
+
+    if driverData.penDefID ~= nil then
+        -- Spawn an alternative weapondef:
+        penDefID = driverData.penDefID
+        timeToLive = timeToLive * driverData.ttlRatio
+        local velRatio = driverData.velRatio
+        vx, vy, vz, vw = vx * velRatio,
+                         vy * velRatio,
+                         vz * velRatio,
+                         vw * velRatio
+        -- Penetrators may or may not be drivers, themselves:
+        penetrator[1] = weaponParams[penDefID] -- nil if not
+    end
+
+    local mx, my, mz, radius
+    if unitID ~= nil then
+        local _
+        _, _, _, mx, my, mz = spGetUnitPosition(unitID, true)
```

Comment:
Why is this split across two lines?

Reply:
I meant to test something with this while I had debugging set up -- left it in. I'm not sure this kind of micro-opt works. Will fix.